1. Basics/task_5/hw5.php

```diff
@@ -0,0 +1,309 @@
+}
+
+/**
+ * Gets input request sample as a string and returns array of the needed request parts
+ * 
+ * @param string $string
+ *
+ * @return array
+ */
+function parseTcpStringAsHttpRequest($string) : array {
+    $strToParse = getStrToParse($string);
+    $strAsArr = cleanArray(explode(PHP_EOL, $strToParse));
+    
+    // parse and delete method and uri
+    $methodAndURI = explode(" ", array_shift($strAsArr));
+    
+    // find body if it is and delete body from the array
+    $body = "";
+    $arrSize = count($strAsArr);
+    
+    if($arrSize > 0) {
+        if(hasBody($string, $strAsArr[$arrSize-1])) {
+            $body = array_pop($strAsArr);
+            $arrSize--;
+        }
+    }
+
+    // parse headers[]
+    $headers = [];
+    $i = 0;
+    
+    // if the request has headers store each header data as
+    // $headers[headerName] = headerValue
+    while($i < $arrSize) {
+        $ind = strpos($strAsArr[$i], ":");
+        $headers[trim(substr($strAsArr[$i], 0, $ind))] = trim(substr($strAsArr[$i], $ind+1));
+        $i++;
+    }
+    
+    return [
+        "method" => trim($methodAndURI[0]),
+        "uri" => trim($methodAndURI[1]),
+        "headers" => $headers,
+        "body" => $body,
+    ];
+}
+
+$http = parseTcpStringAsHttpRequest($contents);
+processHttpRequest($http["method"], $http["uri"], $http["headers"], $http["body"]);
+
+//REQUEST PROCESSNG FUNCTIONS///////////////////////////////////////////////////////////
+
+/**
+ * Gets string as it input and returns a string which let use PHP_EOL as a delimeter
+ * 
+ * @param string $string
+ * 
+ * @return string
+ */
+ function getStrToParse($string) : string {
+    $str = preg_replace("/\r/", "", $string);
+    $str = preg_replace("/\n/", PHP_EOL, $str);
+
+    return $str;
+}
+
+/**
+ * @param string $string
+ * @param string $arrLastItem
+ * 
+ * @return bool
+ */
+function hasBody($string, $arrLastItem) : bool {
+    return preg_match('/Content-Length:/', $string) || !preg_match('/:/', $arrLastItem);
+}
+
+//ARRAYS PROCESSNG FUNCTIONS////////////////////////////////////////////////////////////
+
+/**
+ * Removes items which are empty lines from the array
+ * 
+ * @param array $arr
+ * 
+ * @return array
+ */
+function cleanArray($arr) : array {
+    $resultArr = [];
+    
+    foreach($arr as $item) {
+        if(strlen($item) > 0){
+            $resultArr[]=$item;
+        }
+    }
+
+    return $resultArr;
+}
+
+// RESPONSE MAKING FUNCTIONS///////////////////////////////////////////////////////////
+
+/**
+ * Forms the server response depends on the client request
+ * 
+ * @param string    $requestMethod
+ * @param string    $requestUri
+ * @param array     $requestHeaders
+ * 
+ * @return array
+ */
+function getResponse($requestMethod, $requestUri, $requestHeaders) : array {
+    try {
+        checkRequestMethod($requestMethod);
+        $baseDirectory = getBaseDirectoryByHost($requestHeaders);
+        $filepath = getFilepath($baseDirectory, $requestUri);
```

Comment:
you can remove `$baseDirectory` and pass the function call instead of passing the redundant variable

Reply:
🆗 Done. Now it looks better.